gqcp/include/QCMethod/CC/T2DIIS.hpp

```diff
@@ -0,0 +1,115 @@
+//
+// GQCG-GQCP is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with GQCG-GQCP.  If not, see <http://www.gnu.org/licenses/>.
+
+#pragma once
+
+
+#include "Mathematical/Algorithm/Step.hpp"
+#include "Mathematical/Optimization/Accelerator/DIIS.hpp"
+#include "QCMethod/CC/CCDAmplitudesUpdate.hpp"
+#include "QCMethod/CC/CCDIntermediatesUpdate.hpp"
+#include "QCMethod/CC/CCSDEnvironment.hpp"
+
+#include <algorithm>
+
+
+namespace GQCP {
+
+
+/**
+ *  An iteration step that accelerates the T2 amplitudes based on a DIIS accelerator.
+ * 
+ *  @tparam _Scalar              The scalar type used to represent the T2 amplitudes.
+ */
+template <typename _Scalar>
+class T2DIIS:
+    public Step<CCSDEnvironment<_Scalar>> {
+
+public:
+    // The scalar type used to represent the T2 amplitudes.
+    using Scalar = _Scalar;
+
+    // The type of environment that this iteration step can access.
+    using Environment = CCSDEnvironment<Scalar>;
+
+
+private:
+    // The minimum number of T2 amplitudes that have to be in the subspace before enabling DIIS.
+    size_t minimum_subspace_dimension;
+
+    // The maximum number of T2 amplitues that can be handled by DIIS.
+    size_t maximum_subspace_dimension;
+
+    // The DIIS accelerator.
+    DIIS<Scalar> diis;
+
+
+public:
+    /*
+     *  MARK: Constructors
+     */
+
+    /**
+     *  @param minimum_subspace_dimension       The minimum number of T2 amplitudes that have to be in the subspace before enabling DIIS.
+     *  @param maximum_subspace_dimension       The maximum number of T2 amplitues that can be handled by DIIS.
+     */
+    T2DIIS(const size_t minimum_subspace_dimension = 6, const size_t maximum_subspace_dimension = 6) :
+        minimum_subspace_dimension {minimum_subspace_dimension},
+        maximum_subspace_dimension {maximum_subspace_dimension} {}
+
+
+    /*
+     *  MARK: Conforming to `Step`.
+     */
+
+    /**
+     *  @return A textual description of this algorithmic step.
+     */
+    std::string description() const override {
+        return "Calculate the accelerated T2 amplitudes and place them in the environment by overwriting the previous T2 amplitudes.";
+    }
+
+
+    /**
+     *  Calculate the accelerated T2 amplitudes and place them in the environment by overwriting the previous T2 amplitudes.
+     * 
+     *  @param environment              The environment that acts as a sort of calculation space.
+     */
+    void execute(Environment& environment) override {
+
+        // Don't do anything if the minimum number of T2 amplitude iterations isn't satisfied.
+        if (environment.t2_amplitude_errors.size() < this->minimum_subspace_dimension) {
+            return;
```

Comment:
Why don't we return an error if `t2_amplitude_errors.size() <minimum_subspace_dimension` happens?

Reply:
We only want to _activate_ the DIIS when the minimum subspace dimension has been reached.